backend/uclapi/dashboard/api_applications.py

```diff
@@ -183,3 +207,152 @@ def delete_app(request):
+        return HttpResponseBadRequest(json.dumps({
+            "success": False,
+            "error": "Request is not of method POST"
+        }))
+    try:
+        app_id = request.POST["app_id"]
+        new_callback_url = request.POST["callback_url"]
+        user_id = request.session["user_id"]
+    except KeyError:
+        response = PrettyJsonResponse({
+            "success": False,
+            "message": "Request does not have app_id."
+        })
+        response.status_code = 400
+        return response
+
+    # Check if the new callback URL uses an acceptable protocol
+    # (e.g. HTTP or HTTPS)
+    # The list comprehension will explode the
+    # UCLAPI_OAUTH_CALLBACK_ALLOWED_PROTOCOLS environment
+    # variable, split by semicolons, append each one with :// then
+    # check if the URL starts with it.
+    # If none of them work then bail out.
+    if not any([new_callback_url.startswith(p + "://") for p in os.environ.get(
+            "UCLAPI_OAUTH_CALLBACK_ALLOWED_PROTOCOLS").split(';')]):
+        response = PrettyJsonResponse({
+            "success": False,
+            "message": ("The requested callback URL"
+                        " does not use an acceptable protocol.")
+        })
+        response.status_code = 400
+        return response
+
+    url_data = tldextract.extract(new_callback_url)
+
+    if any(
+        [p == (url_data.domain + "." + url_data.suffix)
+            for p in os.environ.get(
+            "UCLAPI_OAUTH_CALLBACK_DENIED_URLS").split(';')]):
+        response = PrettyJsonResponse({
+            "success": False,
+            "message": ("The requested callback URL"
+                        " is hosted on a banned domain.")
+        })
+        response.status_code = 400
+        return response
+
+    user = get_user_by_id(user_id)
+
+    apps = App.objects.filter(id=app_id, user=user)
+    if len(apps) == 0:
+        response = PrettyJsonResponse({
+            "success": False,
+            "message": "App does not exist."
+        })
+        response.status_code = 400
+        return response
+
+    app = apps[0]
+    app.callback_url = new_callback_url
+    app.save()
+
+    keen_add_event.delay("App callback URL changed", {
+        "appid": app_id,
+        "userid": user.id,
+        "newcallbackurl": new_callback_url
+    })
+
+    return PrettyJsonResponse({
+        "success": True,
+        "message": "Callback URL successfully changed.",
+    })
+
+
+def update_scopes(request):
+    if request.method != "POST":
+        return HttpResponseBadRequest(json.dumps({
+            "success": False,
+            "error": "Request is not of method POST"
+        }))
+
+    try:
+        app_id = request.POST["app_id"]
+        scopes_json = request.POST["scopes"]
+        user_id = request.session["user_id"]
+    except KeyError:
+        response = PrettyJsonResponse({
+            "success": False,
+            "message": "Request does not have app_id."
```

Comment:
message should include that there may not be scopes attached

Reply:
👍